Implement flag metadata

kvnhmn opened the pull request:
Re: [#2682](https://github.com/flipt-io/flipt/issues/2682) (only Backend implementation, no UI changes)

GeorgeMac reviewed (changes requested):
Hey @kvnhmn 🙌 

This is amazing, thank you for taking a run at this.
We would love to help to get this over the line.
The bulk of the PR looks great 🙇 

One thing that would help us would be to leverage the protobuf struct type (instead of the string type).
Really, we would like our old attachment to leverage that too, but its tricky with backwards compat and protobuf.
I added a comment in the review that gives some examples of where we use it in the auth storage layer for metadata there.

Another thing that would help is to update the declarative format.
You can find the definition for Flag here:
https://github.com/flipt-io/flipt/blob/fee220d0a20adfb21686685bef2a1d6c2ff6fc17/internal/ext/common.go#L15-L24

Having Metadata here will allow folks to write it in yaml when either using a declarative backend or importing/exporting via the API.
This will need updating also in import and export:
- import https://github.com/flipt-io/flipt/blob/fee220d0a20adfb21686685bef2a1d6c2ff6fc17/internal/ext/importer.go#L123-L129
- export https://github.com/flipt-io/flipt/blob/fee220d0a20adfb21686685bef2a1d6c2ff6fc17/internal/ext/exporter.go#L129-L135 

Lastly, would be awesome to get at-least a happy path case in the integration tests.
But also, don't want to overload you. This could also be something we do in a subsequent PR.

Thanks for contributing 🙌

kvnhmn commented on rpc/flipt/flipt.proto:
```diff
@@ -110,6 +110,7 @@ message Flag {
   repeated Variant variants = 7;
   string namespace_key = 8;
   FlagType type = 9;
+  string metadata = 10;
```
Of course I can use the Struct 👍🏼

GeorgeMac reviewed (commented):
Thanks for making these changes @kvnhmn . It is looking great.

Unsure if you can see the failures in CI, but thought I would shine a light on some of them.

One of our test suites is seeing a non-nil pointer to an empty metadata:
```
         	            	--- Expected
        	            	+++ Actual
        	            	@@ -16779,3 +16779,16 @@
        	            	  Type: (flipt.FlagType) 0,
        	            	- Metadata: (*structpb.Struct)(<nil>)
        	            	+ Metadata: (*structpb.Struct)({
        	            	+  state: (impl.MessageState) {
        	            	+   NoUnkeyedLiterals: (pragma.NoUnkeyedLiterals) {
        	            	+   },
        	            	+   DoNotCompare: (pragma.DoNotCompare) {
        	            	+   },
        	            	+   DoNotCopy: (pragma.DoNotCopy) {
        	            	+   },
        	            	+   atomicMessageInfo: (*impl.MessageInfo)(<nil>)
        	            	+  },
        	            	+  sizeCache: (int32) 0,
        	            	+  unknownFields: ([]uint8) <nil>,
        	            	+  Fields: (map[string]*structpb.Value) <nil>
        	            	+ })
        	            	 })
```

Which, I think is fine for the API to return this, but either the API needs to change or that test case needs to be updated to ensure it is asserting a non-nil empty struct.

The other thing I am noticing is that the `AFTER` keyword on `ALTER` appears to not be supported by a majority of the relational DB backends.
What is weird is that the SQLite docs says it does not support it, but also, it doesn't fail on migration. Whereas e.g. Postgres fails explicitly during the migrations:
```
 35  : [27.4s] | Error: running migrations: migration failed: syntax error at or near "AFTER" (column 44) in line 1: ALTER TABLE flags ADD COLUMN metadata JSON AFTER enabled;
35  : [27.4s] |  (details: ERROR: syntax error at or near "AFTER" (SQLSTATE 42601))
```

We might need to drop that from the relevant migrations. It seems like maybe MySQL supports it, but not postgres and potentially SQLite just ignores it (so no harm I guess).

kvnhmn commented:
> Thanks for making these changes @kvnhmn . It is looking great.
> 
> Unsure if you can see the failures in CI, but thought I would shine a light on some of them.
> 
> One of our test suites is seeing a non-nil pointer to an empty metadata:
> 
> ```
>          	            	--- Expected
>         	            	+++ Actual
>         	            	@@ -16779,3 +16779,16 @@
>         	            	  Type: (flipt.FlagType) 0,
>         	            	- Metadata: (*structpb.Struct)(<nil>)
>         	            	+ Metadata: (*structpb.Struct)({
>         	            	+  state: (impl.MessageState) {
>         	            	+   NoUnkeyedLiterals: (pragma.NoUnkeyedLiterals) {
>         	            	+   },
>         	            	+   DoNotCompare: (pragma.DoNotCompare) {
>         	            	+   },
>         	            	+   DoNotCopy: (pragma.DoNotCopy) {
>         	            	+   },
>         	            	+   atomicMessageInfo: (*impl.MessageInfo)(<nil>)
>         	            	+  },
>         	            	+  sizeCache: (int32) 0,
>         	            	+  unknownFields: ([]uint8) <nil>,
>         	            	+  Fields: (map[string]*structpb.Value) <nil>
>         	            	+ })
>         	            	 })
> ```
> 
> Which, I think is fine for the API to return this, but either the API needs to change or that test case needs to be updated to ensure it is asserting a non-nil empty struct.
> 
> The other thing I am noticing is that the `AFTER` keyword on `ALTER` appears to not be supported by all of the relational DB backends. What is weird is that the SQLite docs says it does not support it, but also, it doesn't fail on migration. Whereas e.g. Postgres fails explicitly during the migrations:
> 
> ```
>  35  : [27.4s] | Error: running migrations: migration failed: syntax error at or near "AFTER" (column 44) in line 1: ALTER TABLE flags ADD COLUMN metadata JSON AFTER enabled;
> 35  : [27.4s] |  (details: ERROR: syntax error at or near "AFTER" (SQLSTATE 42601))
> ```
> 
> We might need to drop that from the relevant migrations. It seems like maybe MySQL supports it, but not postgres and potentially SQLite just ignores it (so no harm I guess).

Thanks for your tip @GerogeMac. 

I have extended a non-nil empty struct in the integration tests. I also removed it from all migrations, except for the one for MySql, for sqlite I also removed it to be on the safe side, hope that's okay. Also i extended the DEVELOPMENT.md, cause i missed more informations regarding unit-/integration-tests

GeorgeMac commented on rpc/flipt/flipt.proto:
```diff
@@ -110,6 +111,7 @@ message Flag {
   repeated Variant variants = 7;
   string namespace_key = 8;
   FlagType type = 9;
+  google.protobuf.Struct metadata = 10;
```
Im seeing a lot of errors in tests again because the API clients and so on are expecting a non-nil metadata.
I think, accepting nil is something we probably can handle. It might work if you set each of these metadata fields as `optional`.

We could try that and see if it makes the tests happy.

```suggestion
  optional google.protobuf.Struct metadata = 10;
```

markphelps commented on internal/ext/exporter.go:
```diff
@@ -132,6 +132,7 @@ func (e *Exporter) Export(ctx context.Context, encoding Encoding, w io.Writer) e
 					Type:        f.Type.String(),
 					Description: f.Description,
 					Enabled:     f.Enabled,
+					Metadata:    f.Metadata,
```
@kvnhmn the issue with CI looks like it stems from here. `structpb.Struct` marshalls as json using the following code: https://github.com/protocolbuffers/protobuf-go/blob/c33baa8f3a0d35fd5a39e43c22a50a050f707d34/types/known/structpb/struct.pb.go#L226-L228

This looks to me like it will always output `Fields` as a key since it is exported https://github.com/protocolbuffers/protobuf-go/blob/v1.34.2/types/known/structpb/struct.pb.go#L194

The CLI export tests as well as the general export tests are failing because the data thats being added here (metadata) is marshalling as:

```
metadata:
  fields: {}
```

https://github.com/flipt-io/flipt/actions/runs/9863131371/job/27235210359#step:3:623

Which the current tests do not expect.

I think we likely want to fix this by doing the following:

- implementing our own `MarshalJSON` method for the Metadata type (by likely wrapping it in a new type) and check if its empty, and do not marshal if its empty.
- Also we likely want to call the [AsMap](https://github.com/protocolbuffers/protobuf-go/blob/c33baa8f3a0d35fd5a39e43c22a50a050f707d34/types/known/structpb/struct.pb.go#L217-L224) method on it if non-nil, which will prevent the `fields` key from being marshalled and return a plain map

markphelps replied:
cc @erka @GeorgeMac

GeorgeMac replied:
I added an example in a review comment below. I don't think we need a custom marshal json method. I think we just need, as suggested, to make `Metadata` on the `ext/common.go` version of flag to be a `map[string]string`. Then use `AsMap` on export. There is an similar method for adapting a map into a structpb on import too.

Then we should be OK to just let `omitempty` do its thing.

markphelps reviewed (commented):
looking great! commented on failing ci tests

GeorgeMac commented on internal/ext/common.go:
```diff
@@ -13,14 +15,15 @@ type Document struct {
+	Type        string           `yaml:"type,omitempty" json:"type,omitempty"`
+	Description string           `yaml:"description,omitempty" json:"description,omitempty"`
+	Enabled     bool             `yaml:"enabled" json:"enabled"`
+	Metadata    *structpb.Struct `yaml:"metadata,omitempty" json:"metadata,omitempty"`
```
```suggestion
	Metadata   map[string]string `yaml:"metadata,omitempty" json:"metadata,omitempty"`
```

To follow on from @markphelps suggestion. If we just make this a `map[string]string` here we can use `Metadata.AsMap()` when populating this instead.

markphelps reviewed (approved):
Looks great! Thank you both @kvnhmn and @erka !

markphelps commented:
@all-contributors please add @kvnhmn  for code
